File: programers_Lv1/[PCCP 기출문제] 1번.py
# https://school.programmers.co.kr/learn/courses/30/lessons/250137?language=python3
def solution(bandage, health, attacks):
    t, x, y = bandage

    attack_time = []
    damage = []
    for a, b in attacks:
        attack_time.append(a)
        damage.append(b)

    player_health = health
    bonus = 0
    for i in range(1, attack_time[-1] + 1):
        if i in attack_time:
            bonus = 0
            player_health -= damage[attack_time.index(i)]
            if player_health <= 0:
                return -1
            continue
        if player_health < health:
            bonus += 1
            player_health += x
        if bonus == t:
            bonus = 0
            player_health += y
        if player_health > health:
            player_health = health

    return player_health


print(solution([5, 1, 5], 30, [[2, 10], [9, 15], [10, 5], [11, 5]]))

File: programers_Lv1/[PCCP 기출문제] 1번.js
// https://school.programmers.co.kr/learn/courses/30/lessons/250137?language=javascript
function solution(bandage, health, attacks) {
  const [t, x, y] = bandage;

  const attackTime = [];
  const damage = [];

  attacks.forEach((v) => {
    const [a, b] = v;
    attackTime.push(a);
    damage.push(b);
  });

  let playerHealth = health;
  let bonus = 0;

  const arr = Array.from(
    { length: attackTime[attackTime.length - 1] },
    (_, i) => i + 1
  );

  for (v of arr) {
    if (attackTime.includes(v)) {
      bonus = 0;
      playerHealth -= damage[attackTime.indexOf(v)];
      if (playerHealth <= 0) return -1;
      continue;
    }
    if (playerHealth < health) {
      bonus += 1;
      playerHealth += x;
    }
    if (bonus == t) {
      bonus = 0;
      playerHealth += y;
    }
    if (playerHealth > health) {
      playerHealth = health;
    }
  }
  return playerHealth;
}

console.log(
  solution([5, 1, 5], 30, [
    [2, 10],
    [9, 15],
    [10, 5],
    [11, 5],
  ])
);
